src/quicklist.c

```diff
@@ -776,8 +772,31 @@ void quicklistDelEntry(quicklistIter *iter, quicklistEntry *entry) {
 
 /* Replace quicklist entry by 'data' with length 'sz'. */
 void quicklistReplaceEntry(quicklist *quicklist, quicklistEntry *entry,
-                           void *data, int sz) {
+                           void *data, size_t sz) {
+    if(QL_NODE_IS_PLAIN(entry->node)) {
+        if(isLargeElement(sz)) {
+            zfree(entry->node->entry);
+            entry->node->entry = zmalloc(sz);
+            entry->node->sz = sz;
+            memcpy(entry->node->entry, data, sz);
+        } else {
+            quicklistInsertAfter(quicklist, entry, data, sz);
+            __quicklistDelNode(quicklist, entry->node);
+        }
+    } else if(isLargeElement(sz)) {
+        quicklistInsertAfter(quicklist, entry, data, sz);
+
+        if(entry->node->count == 1)
+            __quicklistDelNode(quicklist, entry->node);
+        else
+            ziplistDelete(entry->node->entry, &entry->zi);
+    }
+
     /* quicklistNext() and quicklistIndex() provide an uncompressed node */
+    if(QL_NODE_IS_PLAIN(entry->node) || isLargeElement(sz)) {
+        quicklistCompress(quicklist, entry->node);
+        return;
+    }
```

Comment:
This is just to avoid call `quicklsitCompress` twice?
If that's the case I think we can move `zipstlistReplace` to above ifs(add else).